File: python-algo/leetcode/problem_539.py
"""539. Minimum Time Difference
https://leetcode.com/problems/minimum-time-difference/submissions/
"""
from typing import List


def find_min_difference(time_points: List[str]) -> int:
    def helper(t1: str, t2: str) -> int:
        h1, m1 = t1.split(":")
        h2, m2 = t2.split(":")
        dh, dm = int(h2) - int(h1), int(m2) - int(m1)
        ret = dh * 60 + dm
        return ret if ret < 720 else 1440 - ret

    time_points.sort()
    ans = helper(time_points[0], time_points[-1])
    for i in range(1, len(time_points)):
        ans = min(ans, helper(time_points[i - 1], time_points[i]))
    return ans

File: java-algo/src/main/java/com/leetcode/Problem539.java
package com.leetcode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 539. Minimum Time Difference
 * https://leetcode.com/problems/minimum-time-difference/
 */
public class Problem539 {
    public int findMinDifference(List<String> timePoints) {
        List<Integer> list = new ArrayList<>();
        for (String point : timePoints) {
            String[] segs = point.split(":");
            list.add(Integer.parseInt(segs[0]) * 60 + Integer.parseInt(segs[1]));
        }
        Collections.sort(list);
        int ans = 24 * 60;
        for (int i = 0; i < list.size(); i++) {
            int diff = Math.abs(list.get((i + 1) % list.size()) - list.get(i));
            ans = Math.min(ans, Math.min(diff, 1440 - diff));
        }
        return ans;
    }
}
